### PR TITLE
Add feature to request data from a source that you don't have access to

### PR DESCRIPTION
**Pull request by codazzo**
This PR gives source admins the option to let users request access to a data source.
When this option is enabled, users can see the data source in the "organizations" page and upon opening it are presented with a "Request Access Button".
When the button is clicked, a form is displayed to let the user request access to a source by specifying the organization on behalf of which they are making the request and a message to go with it.
When the request is submitted, an email is sent to the highest-ranking member of the organization (in an order specified by role: 'developer', 'manager', 'founder', 'member').
The email contains a link to the "access" section of the data source which displays the access request with the requester's email address, organization and the message.
Upon accepting, the requesting organization is added to the "allowed organizations" of the source and a confirmation email is sent to the requester.
Upon ignoring, the request is discarded and no longer displayed in the access tab.

**Screenshots**
[source admin tab](https://cloud.githubusercontent.com/assets/167537/22789493/5ec64e68-eee3-11e6-96ed-0c13cf9e9e5c.png)
["request access" button](https://cloud.githubusercontent.com/assets/167537/22789544/82322cfa-eee3-11e6-8317-0ab84520ba60.png)
["request access" form](https://cloud.githubusercontent.com/assets/167537/22789585/ad896148-eee3-11e6-8694-0720aef549a2.png)
["access" tab with pending request](https://cloud.githubusercontent.com/assets/167537/22789621/ccd8af0e-eee3-11e6-8085-3d325e6d831a.png)

Trello link: https://trello.com/c/oOmVIb6k/293-add-feature-to-request-data-from-a-source-that-you-don-t-have-access-to

### CODE REVIEW

File: client/pages/sources/show/overview/map.html [hunk 1]
```diff
@@ -2,29 +2,79 @@
...
+              <p>
+                By clicking "Confirm", the admin of the source will be notified that
+                {{email}}
+                is requesting access.
+              </p>
+
+              <br>
+
+              <a class="call-to-action button js-confirm-access-request">
+                Confirm
```
Comment (opyh) [inline client/pages/sources/show/overview/map.html]: Wording proposal: 'Confirm &amp; send email`

File: client/pages/sources/show/overview/map.js [hunk 2]
```diff
@@ -171,14 +178,106 @@ Template.sources_show_page_map.onCreated(function created() {
...
+
+reactiveVariables.forEach(helper =>
   Template.sources_show_page_map.helpers({
     [helper]() { return Template.instance()[helper].get(); },
   })
 );
 
+Template.sources_show_page_map.helpers({
+  hasAccessToSource() {
+    const source = Sources.findOne({ _id: FlowRouter.getParam('_id') });
```
Comment (opyh) [inline client/pages/sources/show/overview/map.js]: I think `this.getSource()` is available in this context.

File: .meteor/versions [hunk 3]
```diff
@@ -46,6 +46,8 @@ es5-shim@4.6.14_1
 factory@1.0.0
 fastclick@1.0.12
 force-ssl@1.0.12
+fortawesome:fontawesome@4.4.0_1
+fourseven:scss@3.8.1
```
Comment (codazzo) [inline .meteor/versions]: I believe these are added by Bert (https://github.com/themeteorchef/bert/blob/master/package.js#L17-L18). Not sure if it's safe to remove them if we don't use them. I could also look for another notification system with no dependencies (should be an easy drop-in).

File: both/api/source-access-requests/server/invitations/request-access-to-source.js [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    SourceAccessRequests.update(selector, {
+      $set: {
+        requestState: 'sent',
+      },
+    });
+  } catch (error) {
+    SourceAccessRequests.update(selector, {
+      $set: {
+        requestState: 'error',
+        requestError: error,
```
Comment (codazzo) [inline both/api/source-access-requests/server/invitations/request-access-to-source.js]: 👍

File: both/api/source-access-requests/server/methods/ignore.js [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    const sourceAccessRequest = SourceAccessRequests.findOne(requestId);
+    if (!sourceAccessRequest) {
+      throw new Meteor.Error(404, TAPi18n.__('Access Request not found'));
+    }
+
+    const source = Sources.findOne(sourceAccessRequest.sourceId);
+    if (!source) {
+      throw new Meteor.Error(404, TAPi18n.__('Source ${sourceAccessRequest.sourceId} not found'));
```
Comment (codazzo) [inline both/api/source-access-requests/server/methods/ignore.js]: Very good point!

File: client/pages/sources/show/access/access.js [hunk 6]
```diff
@@ -31,18 +35,40 @@ const helpers = {
     }
     return '';
   },
+  checkedIfAccessToSourceCanBeRequested() {
+    const source = Sources.findOne({ _id: FlowRouter.getParam('_id') });
```
Comment (codazzo) [inline client/pages/sources/show/access/access.js]: 👍

File: client/pages/sources/show/access/access.js [hunk 7]
```diff
@@ -31,18 +35,40 @@ const helpers = {
...
+
+    return source.isRequestable ? 'checked' : '';
```
Comment (codazzo) [inline client/pages/sources/show/access/access.js]: Makes sense. I like [`_.get`](https://lodash.com/docs/4.17.4#get) for this (particularly when the property path is long). Ok with using it?

File: client/pages/sources/show/access/access.js [hunk 8]
```diff
@@ -31,18 +35,40 @@ const helpers = {
...
+      sourceId: source._id,
+      requestState: 'sent',
+    }).fetch();
+
+    return requestsArray.map(request => {
+      const requester = Meteor.users.findOne(request.requesterId);
+      const requesterOrganization = Organizations.findOne(request.organizationId).name;
+
+      return Object.assign({}, request, {
+        email: requester.emails[0].address,
```
Comment (codazzo) [inline client/pages/sources/show/access/access.js]: Makes sense, will add them

File: client/pages/sources/show/access/access.js [hunk 9]
```diff
@@ -31,18 +35,40 @@ const helpers = {
...
+        requesterOrganization,
+      });
+    });
+  },
 };
 
 
 Template.sources_show_access_page.events({
-  'change select': function (event) {
+  'change select': (event) => {
```
Comment (codazzo) [inline client/pages/sources/show/access/access.js]: 👍

File: client/pages/sources/show/overview/map.html (hunk 1 above)
Comment (codazzo) [inline client/pages/sources/show/overview/map.html]: 👍

File: client/pages/sources/show/overview/map.js (hunk 2 above)
Comment (codazzo) [inline client/pages/sources/show/overview/map.js]: 👍

File: .meteor/versions (hunk 3 above)
Comment (opyh) [inline .meteor/versions]: Then let's leave them in for now!

File: client/pages/sources/show/access/access.js (hunk 7 above)
Comment (opyh) [inline client/pages/sources/show/access/access.js]: Absolutely!

**Comment by codazzo** [comment]
I've pushed a number of commits to address @hypo's comments. Thanks for the review, merging as discussed.

**Pull request merged**